### DISCUSSION TITLE
GitHub Actions Windows Runtime - Linux Container Mode Docker

### DISCUSSION THREAD

**Discussion by SudharakaP** [community/community#25491] [Actions]
<p>Hi there,</p>
<p>While working on the following issue <a href="https://github.com/hipster-labs/jhipster-daily-builds/issues/59" rel="noopener nofollow ugc">https://github.com/hipster-labs/jhipster-daily-builds/issues/59</a>, we encountered a small problem. We are running a GiHub action on <code>ubuntu-latest</code> (windows-2019). However we need to run Docker within this windows runner but in Linux Container mode. It seems by default Docker runs on Windows container mode. Is there any way of doing this?</p>
<p>Best,<br>
Sudharaka.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/sudharakap">@sudharakap</a>,</p>
<p>You can try this:</p>
<pre><code class="lang-auto">&amp; $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchDaemon
</code></pre>
<p>More details, you can reference to this topic:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://aws1.discourse-cdn.com/docker/optimized/3X/1/0/103abe20a14b522aca1dbb6b11f43265aefb7f0d_2_32x32.png" class="site-icon" width="32" height="32">
      <a href="https://forums.docker.com/t/cli-to-switch-between-linux-and-windows-images/30297" target="_blank" rel="noopener nofollow ugc" title="05:31AM - 27 March 2017">Docker Forums – 27 Mar 17</a>
  </header>
  <article class="onebox-body">
    <img src="https://aws1.discourse-cdn.com/docker/original/3X/4/3/4354118a690f2ccb1acea58f05febcc551357e7d.svg" class="thumbnail onebox-full-image" width="60" height="43">

<h3><a href="https://forums.docker.com/t/cli-to-switch-between-linux-and-windows-images/30297" target="_blank" rel="noopener nofollow ugc">CLI to switch between linux and windows images</a></h3>

<p>I can switch between Linux and Windows images using the docker task tray without issue.  My question is in regards to the command line.  I’ve read and searched the docs but I don’t think I’m looking for the right thing.</p>

  <p><span class="label1">Reading time: 1 mins 🕑</span>
    <span class="label2">Likes: 6 ❤</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by SudharakaP** [comment]
<p><a class="mention" href="/brightran">@brightran</a>: Thanks much for the response; but it seems that when I try to run that it gives me the following error;</p>
<pre><code class="lang-auto">Run &amp; $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchDaemon
&amp; : The term 'C:\Program Files\Docker\Docker\DockerCli.exe' is not recognized as the name of a cmdlet, function, 
script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is 
correct and try again.
At D:\a\_temp\d38dac67-e8c7-41bc-8f8b-f1d8788dfd5b.ps1:2 char:3
+ &amp; $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchDaemon
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Program File...r\DockerCli.exe:String) [], ParentContainsErrorRecord 
   Exception
    + FullyQualifiedErrorId : CommandNotFoundException
</code></pre>
<p>I think there does not exist <code>DockerCli.exe</code> in these VMs. :frowning: Here’s the link to the CI run; <a href="https://github.com/SudharakaP/jhipster-daily-builds/runs/1213300687?check_suite_focus=true" rel="noopener nofollow ugc">https://github.com/SudharakaP/jhipster-daily-builds/runs/1213300687?check_suite_focus=true</a></p>

**Comment by chrispat** [comment, marked as answer]
<p>The Windows VMs are not enabled for nested virtualization so you won’t be able to run a linux container on the Windows VM.  We will not be able to enable that feature due to the VMs already being nested on level deep and they hypervisor does not support multiple levels of nested virtualization.</p>

**Reply by pseusys** [reply]
Hello, @chrispat!
Some years have passed since your response and now the installation documentation of Docker Desktop specifies WSL2 as a [possible alternative](https://docs.docker.com/desktop/setup/install/windows-install/#system-requirements) to Hyper-V. WSL2, in turn, does require virtualisation, however there is WSL2 [already installed](https://github.com/actions/runner-images/blob/main/images/windows/Windows2025-Readme.md) on the latest GitHub Actions Windows runner image (not even mentioning several custom actions providing it on the previous runner versions). Does this somehow change the game? Maybe it would be possible to run docker executable from-inside WSL2? Or is there a deeper virtualisation requirement underneath that **does** allow running WSL2 but **does not** allow docker? I am just curious how it works at this point :)

**Reply by roger-zhangg** [reply]
Exact same question as above. @pseusys any luck with WSL2?

**Reply by pseusys** [reply]
> any luck with WSL2?

No, no luck. As far as I remember, the latest thing that I learned about the topic is that there are two different versions of Docker for Windows and Windows Server (which is used by the Actions images). The one for the Server is not really integrated with WSL yet. Still, you can for sure run WSL and Docker in parallel on a Linux machine, if that helps.

**Comment by SudharakaP** [comment]
<p><a class="mention" href="/chrispat">@chrispat</a> : Thanks much for the information. Appreciate it. :+1:</p>
